crate_universe/src/metadata.rs

```diff
@@ -319,6 +324,127 @@ impl VendorGenerator {
     }
 }
 
+/// A generate which computes per-platform feature sets.
+pub struct FeatureGenerator {
+    /// The path to a `cargo` binary
+    cargo_bin: PathBuf,
+
+    /// The path to a `rustc` binary
+    rustc_bin: PathBuf,
+}
+
+impl FeatureGenerator {
+    pub fn new(cargo_bin: PathBuf, rustc_bin: PathBuf) -> Self {
+        Self {
+            cargo_bin,
+            rustc_bin,
+        }
+    }
+
+    /// Computes the set of enabled features for each target triplet for each crate.
+    pub fn generate(
+        &self,
+        manifest_path: &Path,
+        platform_triples: &BTreeSet<String>,
+    ) -> Result<BTreeMap<CrateId, SelectList<String>>> {
+        let manifest_dir = manifest_path.parent().unwrap();
+        let mut crate_features = BTreeMap::<CrateId, BTreeMap<String, BTreeSet<String>>>::new();
+        for target in platform_triples {
+            let output = Command::new(&self.cargo_bin)
+                .current_dir(manifest_dir)
+                .arg("tree")
+                .arg("--locked")
+                .arg("--manifest-path")
+                .arg(manifest_path)
+                .arg("--prefix=none")
+                .arg("--format=|{p}|{f}|")
+                .arg("--color=never")
+                .arg("--target")
+                .arg(target)
+                .env("RUSTC", &self.rustc_bin)
+                .output()
+                .context(format!(
+                    "Error running cargo to compute features for target '{}', manifest path '{}'",
+                    target,
+                    manifest_path.display()
+                ))?;
+            if !output.status.success() {
+                eprintln!("{}", String::from_utf8_lossy(&output.stdout));
+                eprintln!("{}", String::from_utf8_lossy(&output.stderr));
+                bail!(format!("Failed to run cargo tree: {}", output.status))
+            }
+            for line in output.stdout.lines() {
```

Comment:
Would it be possible to split the logic below into it's own function so it could be tested separately? I think it'd be helpful to have some small examples of what the output is expected to be from this cargo command live somewhere in the repo in plain text

Reply:
Done... it's a little awkward since `output.stdout.lines()` is an iterator over `Result<String>`. Added a fairly simple test case to go with it.